src/transform/src/canonicalization.rs

```diff
@@ -21,3 +21,134 @@ mod topk_elision;
+        level = "debug",
+        fields(path.segment = "reduce_scalars")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        ctx: &mut TransformCtx,
+    ) -> Result<(), crate::TransformError> {
+        let mut builder = DerivedBuilder::new(ctx.features);
+        builder.require(RelationType);
+        let derived = builder.visit(&*relation);
+
+        // Descend the AST, removing `Threshold` operators whose inputs are non-negative.
+        let mut todo = vec![(&mut *relation, derived.as_view())];
+        while let Some((expr, view)) = todo.pop() {
+            match expr {
+                MirRelationExpr::Constant { .. }
+                | MirRelationExpr::Get { .. }
+                | MirRelationExpr::Let { .. }
+                | MirRelationExpr::LetRec { .. }
+                | MirRelationExpr::Project { .. }
+                | MirRelationExpr::Union { .. }
+                | MirRelationExpr::Threshold { .. }
+                | MirRelationExpr::Negate { .. } => {
+                    // No expressions to reduce
+                }
+                MirRelationExpr::ArrangeBy { .. } => {
+                    // Has expressions, but we aren't brave enough to reduce these yet.
+                }
+                MirRelationExpr::Filter { predicates, .. } => {
+                    let input_type = view
+                        .last_child()
+                        .value::<RelationType>()
+                        .expect("RelationType required")
+                        .as_ref()
+                        .unwrap();
+                    for predicate in predicates.iter_mut() {
+                        predicate.reduce(input_type);
+                    }
+                    predicates.retain(|p| !p.is_literal_true());
+                }
+                MirRelationExpr::FlatMap { exprs, .. } => {
+                    let input_type = view
+                        .last_child()
+                        .value::<RelationType>()
+                        .expect("RelationType required")
+                        .as_ref()
+                        .unwrap();
+                    for expr in exprs.iter_mut() {
+                        expr.reduce(input_type);
+                    }
+                }
+                MirRelationExpr::Map { scalars, .. } => {
+                    // Use the output type, to incorporate the types of `scalars` as they land.
+                    let output_type = view
+                        .value::<RelationType>()
+                        .expect("RelationType required")
+                        .as_ref()
+                        .unwrap();
+                    let input_arity = output_type.len() - scalars.len();
+                    for (index, scalar) in scalars.iter_mut().enumerate() {
+                        scalar.reduce(&output_type[..input_arity + index]);
+                    }
+                }
+                MirRelationExpr::Join { equivalences, .. } => {
```

Comment:
Do we want to maybe also go for the exprs inside `JoinImplementation`?

Reply:
I don't know; they have the same vibes to me as the expressions in `ArrangeBy`: they aren't semantically load-bearing, and they don't have clearly documented invariants other than (afaict) "don't touch!".

Reply:
But could certainly add comments reflecting whatever conclusion we reach about them. I'm not certain what happens here if we optimize them and then e.g. fail to find a match in a child `ArrangeBy`. Perhaps we could reduce both of the expressions, but again .. I'm not certain about the implied contracts between these operators.

Reply:
🌶️ : we should remove `JoinImplementation` from MIR and it should only exist in more explicitly planned form in LIR. The join implementation transform currently has the effect that it identifies arrangements we might need to create, and we can still do that in "physical planning", perhaps, .. but I don't know that MIR wants to be in the business of join implementation strategies (and mostly needs to remember to un-set them if a transform changes anything at all about the surrounding join).

Reply:
Yeah, moving `JoinImplementation` into the MIR-to-LIR lowering came up a few times. I did an experiment once where I simply put `JoinImplementation` at the very end of the MIR pipeline, and there were regressions, but not too bad ones if I remember correctly.